crates/turbopack-core/src/resolve/mod.rs

```diff
@@ -650,16 +779,78 @@ impl ResolveResult {
         Ok(Vc::cell(
             this.primary
                 .iter()
-                .filter_map(|item| {
-                    if let ResolveResultItem::Source(a) = item {
-                        Some(*a)
+                .filter_map(|(_, item)| {
+                    if let &ResolveResultItem::Source(a) = item {
+                        Some(a)
                     } else {
                         None
                     }
                 })
                 .collect(),
         ))
     }
+
+    #[turbo_tasks::function]
+    pub async fn with_replaced_request_key(
+        self: Vc<Self>,
+        old_request_key: String,
+        request_key: Value<RequestKey>,
+    ) -> Result<Vc<Self>> {
+        let this = self.await?;
+        let request_key = request_key.into_value();
+        let new_primary = this
+            .primary
+            .iter()
+            .filter_map(|(k, v)| {
+                if let Some(k) = &k.request {
+                    if let Some(remaining) = k.strip_prefix(&old_request_key) {
+                        Some((
+                            RequestKey {
+                                request: request_key
+                                    .request
+                                    .as_ref()
+                                    .map(|r| format!("{}{}", r, remaining)),
+                                conditions: request_key.conditions.clone(),
+                            },
+                            v.clone(),
+                        ))
+                    } else {
+                        None
+                    }
+                } else {
+                    Some((request_key.clone(), v.clone()))
+                }
+            })
+            .collect();
+        Ok(ResolveResult {
+            primary: new_primary,
+            affecting_sources: this.affecting_sources.clone(),
+        }
+        .into())
+    }
+
+    #[turbo_tasks::function]
+    pub async fn with_request(self: Vc<Self>, request: String) -> Result<Vc<Self>> {
+        let this = self.await?;
+        let new_primary = this
+            .primary
+            .iter()
+            .map(|(k, v)| {
+                (
+                    RequestKey {
+                        request: Some(request.clone()),
+                        conditions: k.conditions.clone(),
```

Comment:
It's expected that this one keeps the old conditions, but the one above does not? (document if it is)